### DISCUSSION TITLE
Using Python API to Extract Part Fields

### DISCUSSION THREAD

**Discussion by palagap** [inventree/InvenTree#7045] [Q&A]
Hi there,

I'm trying to use the InvenTree Python API to extract certain fields for parts. For instance, in the screenshot below, I would like to be able to extract the values of fields such as Location, Purchase Order, and Purchase Price. 

<img width="1659" alt="Screenshot InvenTree Part" src="https://github.com/inventree/InvenTree/assets/40808444/7160ea21-033f-4232-ae81-7301578c24cb">

I am looking through both the general Python API documentation (https://docs.inventree.org/en/latest/api/python/python/) and the `api-doc`. Using the Server Address and Token, I'm able to connect to the InvenTree Server. e.g. through the command:

`api = InvenTreeAPI(SERVER_ADDRESS, token=MY_TOKEN)`

However, I am not able to view any specific fields related to each part, or find out the correct primary key of a part of interest. For example, when attempting to print out the result of the`PartsCategory` command, I get a description of the class (e.g. `<class 'inventree.part.PartCategory'><pk=[PN]>`) rather than the part itself.

Do you have any suggestions on how I can access specific fields of a part of interest?

Thanks.

**Comment by SchrodingersGat** [comment]
Uh, yeah, you are printing a list of class objects. If you [read the documentation](https://docs.inventree.org/en/stable/api/python/examples/#updating-attributes) you will see that you can access these attributes by name:

```python
for category in PartCategory.list(api):
    print("Name:", category.name)
```

**Reply by palagap** [reply]
Thanks, I can now view the same attributes (e.g. name, description, minimum_stock) as in the examples for different parts. For attributes not listed in the examples, (e.g. purchase order, purchase price, etc.), which file(s) should I look through to find them? I've looked through both `api.py` and `part.py`, in the documentation, but don't see any references to those particular attributes. I've tried `part.purchase_price` or `part.purchase_order`, but those aren't valid attributes.

**Reply by SchrodingersGat** [reply]
You can look at the data coming from the API itself to see what fields are available in Python.

**Reply by palagap** [reply]
I'll do that, thanks.

**Comment by miggland** [comment]
In Python, you can use the the following to find available fields:

```
p = PurchaseOrder(api, pk=12)
p.fieldNames(api)
p.fields(api)
p._data
```

**Reply by palagap** [reply]
Got it, thanks for the info.
